### PR TITLE
Adds filtering to caption list

### PR DESCRIPTION
**Pull request by tehblasian**

### CODE REVIEW

**Comment by tehblasian** [comment]
There is a bug in the branch, do not review yet

**Comment by tehblasian** [comment]
Bug is fixed

File: frontend/src/util/apiUtil.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
+import Cookies from 'universal-cookie';
+import axios from 'axios';
+
+export const getToken = () => {
```
Comment (justin-cotarla) [inline frontend/src/util/apiUtil.js]: This is already in AuthUtil.js

File: frontend/src/pages/ProfilePage.js [hunk 2]
```diff
@@ -27,24 +25,25 @@ class ProfilePage extends React.Component {
     }
 
     componentDidMount(){
-        const token = this.getToken();
+        const token = getToken();
```
Comment (justin-cotarla) [inline frontend/src/pages/ProfilePage.js]: Token is available in the props

File: frontend/src/pages/ProfilePage.js [hunk 3]
```diff
@@ -168,16 +110,15 @@ class ProfilePage extends React.Component {
                         {
                             selectedView === views[0]
                             && (
-                                ( captions.length === 0 
-                                    && <h1 className="header-section">There aren't any captions to see here :(</h1> )
-                                || <CaptionList 
-                                        captions={captions}
+                                <CaptionList
```
Comment (justin-cotarla) [inline frontend/src/pages/ProfilePage.js:109]: Pass one function only

File: frontend/src/pages/ProfilePage.js [hunk 4]
```diff
@@ -168,16 +110,15 @@ class ProfilePage extends React.Component {
...
                             && (
-                                ( captions.length === 0 
-                                    && <h1 className="header-section">There aren't any captions to see here :(</h1> )
-                                || <CaptionList 
-                                        captions={captions}
+                                <CaptionList 
+                                        fetchCaptions={() => fetchUserCaptions(profileUser.id, token)}
+                                        getFilteredCaptions={filter => getFilteredCaptionsUser(profileUser.id, filter, token)}
                                         showSubmittedBy={false} 
+                                        showCount={false}
```
Comment (justin-cotarla) [inline frontend/src/pages/ProfilePage.js]: Implement this later to show the total caption count

File: frontend/src/util/apiUtil.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    const cookies = new Cookies();
+    const token = cookies.get('token');
+    return token;
+}
+
+export const fetchCaptionsByMomentId = (momentid, token) => {
```
Comment (justin-cotarla) [inline frontend/src/util/apiUtil.js]: Export a single `fetchCaptions` function that takes several params and internally handles building the query.

File: frontend/src/util/apiUtil.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        'Authorization': `Bearer ${token}` 
+    };
+    return axios({
+        method: 'get',
+        url: `http://${process.env.REACT_APP_IP}/api/captions?user-id=${id}`,
+        header: token ? headers : {}
+    });
+}
+
+export const fetchUserMoments = (id, token) => {
```
Comment (justin-cotarla) [inline frontend/src/util/apiUtil.js]: Export a single `fetchMoments` function that takes several params and internally handles building the query.

File: frontend/src/pages/MomentViewPage.js [hunk 7]
```diff
@@ -7,17 +7,17 @@ import CaptionCreatorForm from '../components/CaptionCreatorForm';
 import CaptionList from '../components/CaptionList';
 import NavBar from '../components/NavBar';
 
-import Header from '../components/Header';
 import Loading from '../components/Loading';
 import ErrorGraphic from '../components/ErrorGraphic';
 
+import { fetchCaptionsByMomentId, getFilteredCaptionsByMoment } from '../util/apiUtil';
```
Comment (justin-cotarla) [inline frontend/src/pages/MomentViewPage.js]: Import one function only

File: frontend/src/pages/MomentViewPage.js [hunk 8]
```diff
@@ -54,45 +54,17 @@ class MomentViewPage extends Component{
...
-    // A callback to update the caption list after a vote or acception/rejection
-    onCaptionUpdate = (newcaption) => {
-        this.setState({
-            captions: this.state.captions.map(caption => {
-                if(newcaption.caption_id === caption.caption_id){
-                    return newcaption;
-                }
-                return caption;
-            }),
+    onCaptionSubmit = (momentId) => {
```
Comment (justin-cotarla) [inline frontend/src/pages/MomentViewPage.js]: Don't leak CaptionList functionality here, rather just force it to re-render.

File: frontend/src/components/CaptionList.js [hunk 9]
```diff
@@ -1,25 +1,92 @@
...
+                loading: false,
+            });
+        });
+    }
+
+    onFilterChange = (selectedFilter) => {
+        const currentFilter = this.state.selectedFilter;
+        if (selectedFilter !== currentFilter) {
+            this.setState({ 
+                scrolled: true,
```
Comment (justin-cotarla) [inline frontend/src/components/CaptionList.js]: All scrolling should be handled here, not in caption

File: frontend/src/components/CaptionList.js [hunk 10]
```diff
@@ -1,25 +1,92 @@
...
+        this.props.fetchCaptions()
+        .then(response => {
+            const { captions } = response.data;
+            this.setState({ 
+                captions,
+                loading: false,
+            })
+        })
+        .catch(error => {
+            console.log(error);
```
Comment (justin-cotarla) [inline frontend/src/components/CaptionList.js]: Don't leave these here

File: frontend/src/components/CaptionList.js [hunk 11]
```diff
@@ -1,25 +1,92 @@
...
+                }
+                <CaptionFilter selectedFilter={selectedFilter} onFilterChange={this.onFilterChange}/>
+                <ul>
+                    { 
+                        captions.map(caption => {
+                            return <li key={caption.caption_id} style={isLinkedToMoment ? {cursor: 'pointer'} : {}}>
                                 <ConditionalWrap
                                     condition={isLinkedToMoment}
-                                    wrap={children => <Link className="linked-caption" to={{pathname: `/moment/${caption.moment_id}`, state: { scrollTo: caption.caption_id}}}>{children}</Link>}>
+                                    wrap={children => <Link className="linked-caption" to={{pathname: `/moment/${caption.moment_id}`, state: { scrollTo: caption.caption_id}}}>{children}</Link>}
```
Comment (justin-cotarla) [inline frontend/src/components/CaptionList.js:102]: Break this up

File: frontend/src/components/CaptionFilter.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
+import React from 'react';
+import '../styles/CaptionFilter.css';
+
+const CaptionFilter = props => {
+    const filters = ['Recent', 'Oldest', 'Top', 'Worst', 'Accepted', 'Rejected'];
```
Comment (justin-cotarla) [inline frontend/src/components/CaptionFilter.js]: Have these defined as constants elsewhere

**Review by justin-cotarla — CHANGES_REQUESTED** [review]

File: frontend/src/util/apiUtil.js [hunk 13]
```diff
@@ -1,21 +1,33 @@
...
-    const cookies = new Cookies();
-    const token = cookies.get('token');
-    return token;
+export const captionRequestTypes = {
+    BY_MOMENT: 'BY_MOMENT',
+    BY_USER: 'BY_USER',
+    BY_MOMENT_AND_USER: 'BY_MOMENT_AND_USER',
 }
+export const captionFilters = ['Recent', 'Oldest', 'Top', 'Worst', 'Accepted', 'Rejected'];
+export const fetchCaptions = ({ token, type, filter, momentId, userId }) => {
```
Comment (justin-cotarla) [inline frontend/src/util/apiUtil.js]: I would move this out into something like `getRequestTypeQuery` or something, and give it a similar implementation to `getFilterQuery`. That way this part would remain modular.

**Review by justin-cotarla — CHANGES_REQUESTED** [review]

**Review by justin-cotarla — APPROVED** [review]
:eyes: good, will make fetching captions great again.

**Pull request merged**